File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        if nums is None:
            return res

        def combinationSumBuilder(index,cur,total):
            if total == target:
                res.append(cur[:])
                return

            if total > target:
                return
            
            for i in range(index,len(nums)):
                cur.append(nums[i])
                combinationSumBuilder(i,cur,total + nums[i])
                cur.pop()
        combinationSumBuilder(0,[],0)
        return res;
        

File: Data Structures & Algorithms/subsets/submission-3.cs
public class Solution {
    public List<List<int>> Subsets(int[] nums) {
        List<List<int>> res = new List<List<int>>();
        if (nums == null) {
            return res;
        }

        generateSubset(0,res,nums,new List<int>());
        return res;
    }

    static void generateSubset(int start, List<List<int>> res,
    int[] nums, List<int> cur) {
        res.Add(new List<int>(cur));

        for (int i = start; i < nums.Length; i++) {
            cur.Add(nums[i]);
            generateSubset(i+1,res,nums,cur);
            cur.RemoveAt(cur.Count - 1);

        }
    }
}
